edit/auth/azuread.vue

```diff
@@ -0,0 +1,268 @@
+    authEndpoint:  `https://login.microsoftonline.com/${ TENANT_ID_TOKEN }/oauth2/authorize`,
+  },
+  china: {
+    endpoint:      'https://login.chinacloudapi.cn/',
+    graphEndpoint: 'https://graph.chinacloudapi.cn/',
+    tokenEndpoint: `https://login.chinacloudapi.cn/${ TENANT_ID_TOKEN }/oauth2/token`,
+    authEndpoint:  `https://login.chinacloudapi.cn/${ TENANT_ID_TOKEN }/oauth2/authorize`,
+  },
+  custom: {
+    endpoint:      'https://login.microsoftonline.com/',
+    graphEndpoint: '',
+    tokenEndpoint: '',
+    authEndpoint:  '',
+  }
+};
+
+export default {
+  components: {
+    Loading,
+    CruResource,
+    InfoBox,
+    RadioGroup,
+    LabeledInput,
+    Banner,
+    CopyToClipboardText,
+    AllowedPrincipals,
+    AsyncButton
+  },
+
+  mixins: [CreateEditView, Auth],
+
+  async fetch() {
+    await this.reloadModel();
+
+    this.model = await this.$store.dispatch(`rancher/clone`, { resource: this.originalModel });
+  },
+
+  data() {
+    return {
+      model:         null,
+      errors:        null,
+      endpoint:      'standard'
+    };
+  },
+
+  computed: {
+    baseUrl() {
+      return `${ this.model.tls ? 'https://' : 'http://' }${ this.model.hostname }`;
+    },
+
+    tArgs() {
+      return {
+        baseUrl:   this.baseUrl,
+        provider:  this.displayName,
+        username:  this.principal.loginName || this.principal.name,
+      };
+    },
+
+    NAME() {
+      return NAME;
+    },
+
+    replyUrl() {
+      return `${ window.location.origin }/verify-auth-azure`;
+    },
+    tenantId() {
+      return this.model?.tenantId;
+    },
+    toSave() {
+      return {
+        ...this.model,
+        enabled:           true,
+        description:       'Enable AzureAD',
+        editRedirectUrl: url => `${ url }&response_type=code&response_mode=query`
+      };
+    },
+  },
+
+  watch: {
+    endpoint(value) {
+      this.setEndpoints(value);
+    },
+    tenantId() {
+      this.setEndpoints(this.endpoint);
+    },
+    model() {
+      this.model.accessMode = this.model.accessMode || 'unrestricted';
+      this.model.rancherUrl = this.model.rancherUrl || this.replyUrl;
+      this.setEndpoints(this.endpoint);
+    }
+  },
+
+  methods: {
+    setEndpoints(endpoint) {
+      Object.keys(ENDPOINT_MAPPING[endpoint]).forEach((key) => {
+        this.model[key] = ENDPOINT_MAPPING[endpoint][key].replace(TENANT_ID_TOKEN, this.model.tenantId);
+      });
+    },
+
+    async saveOverride() {
+      // I'm using this workaround because we have some very odd stuff happening with mixins/auth. Mixins auth is actually replacing the model and originalModel with
+      // a norman object. The norman object however doesn't have applicationSecret as part of it's schema which causes the field to get stripped after each reloadModel.
+      // If there's a better way to resolve this we should.
+      const applicationSecret = this.model.applicationSecret || this.value.applicationSecret;

```

Comment:
This is a feature of norman, the secret fields are "write only" and do not come back in the response or subsequent get.

As an alternate you could bind the input to a data field and copy that into the model before saving.

Reply:
I'll move the secret into the data field. However, this is the part I had questions about yesterday.

Why are we even overriding the model/originalModel in reloadModel() to be norman object? When I switched the norman objects to steve objects things failed. I decided to ask during our meeting yesterday so I haven't dug any deeper.

Reply:
The way I did github was the view page works off of the Steve crds; the edit off the Norman resource.  So edit just has the Norman object as it's model/originalModel.